internal/app/referenceclient/wire_tracer.go

```diff
@@ -0,0 +1,159 @@
+func (w *WireWrapper) Unlock() {
+	w.mtx.Unlock()
+}
+
+// Set sets the wire details. Note that calls to Set should always be
+// preceded by calls to Lock.
+func (w *WireWrapper) Set(details *WireDetails) {
+	w.val = details
+}
+
+// WithWireCapture returns a new context which will contain wire details during
+// a roundtrip.
+func WithWireCapture(ctx context.Context) (context.Context, *WireWrapper) {
+	wrapper := &WireWrapper{}
+	ctx = context.WithValue(ctx, wireKey{}, wrapper)
+	return ctx, wrapper
+}
+
+type WireTracer struct {
+	tracer *tracer.Tracer
+}
+
+// Complete intercepts the Complete call for a tracer, extracting wire details
+// from the passed trace. The wire details will be stored in the context acquired byte
+// WithWireCapture and can be retrieved via WireWrapper.Get().
+func (t *WireTracer) Complete(trace tracer.Trace) {
+	wrapper, ok := trace.Request.Context().Value(wireKey{}).(*WireWrapper)
+	// Lock the mutex on the wrapper so that the client implementation isn't
+	// reading the wire details before we get a chance to populate them here.
+	wrapper.Lock()
+	defer wrapper.Unlock()
+	if ok { //nolint:nestif
+		if trace.Response != nil {
+			statusCode := int32(trace.Response.StatusCode)
+
+			var jsonRaw structpb.Struct
+			contentType := trace.Response.Header.Get("content-type")
+			if contentType == "application/json" {
+				if statusCode != 200 {
+					// If this is a unary request, then use the entire response body
+					// as the wire error details.
+					body, err := io.ReadAll(trace.Response.Body)
+					if err != nil {
+						return
+					}
+					if err := protojson.Unmarshal(body, &jsonRaw); err != nil {
+						return
+					}
+				}
+			} else if strings.HasPrefix(contentType, "application/connect+") {
+				type endStreamError struct {
+					Error json.RawMessage `json:"error"`
+				}
+				// If this is a streaming request, then look through the trace events
+				// for the ResponseBodyEndStream event and parse its content into an
+				// endStreamError to see if there are any error details.
+				for _, ev := range trace.Events {
+					switch eventType := ev.(type) {
+					case *tracer.ResponseBodyEndStream:
+						var endStream endStreamError
+						if err := json.Unmarshal([]byte(eventType.Content), &endStream); err != nil {
+							return
+						}
+						if err := protojson.Unmarshal(endStream.Error, &jsonRaw); err != nil {
+							return
+						}
+					default:
+						// Do nothing
+					}
+				}
+			}
+
+			wire := &WireDetails{
+				StatusCode:      statusCode,
+				Trailers:        internal.ConvertToProtoHeader(trace.Response.Trailer),
+				ConnectErrorRaw: &jsonRaw,
+			}
+
+			wrapper.Set(wire)
+		}
+	}
+
+	if t != nil {
+		t.tracer.Complete(trace)
+	}
+}
+
+// NewWireTracer returns a new wire tracer with the given tracer.
+// If trace is nil, all tracer operations will be bypassed.
+func NewWireTracer(trace *tracer.Tracer) *WireTracer {
+	return &WireTracer{
+		tracer: trace,
+	}
+}
```

Comment:
For these things that are only used from one place and are just implementation details of `main`, I'd forego the ceremony of exported types, exported factory functions, etc. So all of these types and functions can be unexported/lowercase, and trivial stuff that is used exactly once, like this function, could be inlined.